File: HTE/jars/src/data/hullmods/drgPDspec.java
package data.hullmods;

import com.fs.starfarer.api.combat.BaseHullMod;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.ShipAPI.HullSize;

public class drgPDspec extends BaseHullMod {

    public static final int 
        PD_RANGE_BONUS = 300, 
        THRESHOLD_AT = 1200;
    // you might be wondering why I use floats some times and ints other times
    // don't worry, I wonder too


    @Override
	public void applyEffectsBeforeShipCreation(HullSize hullSize, MutableShipStatsAPI stats, String id) {
        stats.getNonBeamPDWeaponRangeBonus().modifyFlat(id, PD_RANGE_BONUS);
        stats.getBeamPDWeaponRangeBonus().modifyFlat(id, PD_RANGE_BONUS);
        stats.getWeaponRangeThreshold().modifyFlat(id, 1200);
        stats.getWeaponRangeMultPastThreshold().modifyMult(id, 0f);
    }

    @Override
    public String getDescriptionParam(int index, HullSize hullSize, ShipAPI ship) {
        if (index == 0)
            return Integer.toString(PD_RANGE_BONUS);
        if (index == 1)
            return Integer.toString(THRESHOLD_AT);
        return null;
    }

    @Override
    public boolean affectsOPCosts() {
        return true;
    }

}

File: src/data/hullmods/drgPDspec.kt
package data.hullmods

import com.fs.starfarer.api.combat.BaseHullMod
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.ShipAPI.HullSize
import org.magiclib.util.MagicIncompatibleHullmods

class drgPDspec : BaseHullMod()
{
    // the float/int stuff is to make it display as ints without too much trouble
    override fun applyEffectsBeforeShipCreation(hullSize: HullSize, stats: MutableShipStatsAPI, id: String)
    {
        stats.nonBeamPDWeaponRangeBonus.modifyFlat(id, PD_RANGE_BONUS.toFloat())
        stats.beamPDWeaponRangeBonus.modifyFlat(id, PD_RANGE_BONUS.toFloat())
        stats.weaponRangeThreshold.modifyFlat(id, THRESHOLD_AT.toFloat())
        stats.weaponRangeMultPastThreshold.modifyMult(id, 0f)
    }

    override fun getDescriptionParam(index: Int, hullSize: HullSize, ship: ShipAPI): String?
    {
        if (index == 0) return PD_RANGE_BONUS.toString()
        return if (index == 1) THRESHOLD_AT.toString() else null
    }

    override fun applyEffectsAfterShipCreation(ship: ShipAPI, id: String?)
    {
        if (ship.variant.hullMods.contains("safetyoverrides"))
        {
            MagicIncompatibleHullmods.removeHullmodWithWarning(
                ship.variant,
                "safetyoverrides",
                "drg_pdspec"
            )
        }
    }

    override fun affectsOPCosts(): Boolean
    {
        return true
    }

    companion object
    {
        const val PD_RANGE_BONUS = 300
        const val THRESHOLD_AT = 1200
    }
}